File: Python/Singly_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.ref = None


class LinkedList:
    def __init__(self):
        self.head = None

    def display(self):
        if self.head is None:
            print("Linked list is empty")
        else:
            n = self.head
            while n is not None:
                print(n.data, "->", end=" ")
                n = n.ref
            print()

    def add_begin(self, data):
        new_node = Node(data)
        if self.head is not None:
            new_node.ref = self.head
        self.head = new_node

    def add_end(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            n = self.head
            while n.ref is not None:
                n = n.ref
            n.ref = new_node

    def add_after(self, data, value):
        if self.head is None:
            print("ll is empty")
        else:
            n = self.head
            while n is not None:
                if n.data == value:
                    break
                n = n.ref
            if n is None:
                print("value not found")
            else:
                new_node = Node(data)
                if n.ref is not None:
                    new_node.ref = n.ref
                n.ref = new_node

    def add_before(self, data, value):
        if self.head is None:
            print("linked list is empty")
        elif self.head.data == value:
            new_node = Node(data)
            new_node.ref = self.head
            self.head = new_node
        else:
            n = self.head
            while n.ref is not None:
                if n.ref.data == value:
                    break
                n = n.ref
            if n.ref is None:
                print("not found")
            else:
                new_node = Node(data)
                if n.ref:
                    new_node.ref = n.ref
                n.ref = new_node

    def delete_begin(self):
        if self.head is None:
            print("linked list is empty ")
        else:
            if self.head.ref is not None:
                self.head = self.head.ref
            else:
                self.head = None

    def delete_end(self):
        if self.head is None:
            print("Linked list is empty")
        elif self.head.ref is None:
            self.head = None
        else:
            n = self.head
            while n.ref.ref is not None:
                n = n.ref
            n.ref = None

    def delete_value(self, value):
        if self.head is None:
            print("Linked list is empty")
        elif self.head.data == value:
            if self.head.ref:
                self.head = self.head.ref
            else:
                self.head = None
        else:
            n = self.head
            while n.ref is not None:
                if n.ref.data == value:
                    break
                n = n.ref
            if n.ref is None:
                print("value not found")
            else:
                n.ref = n.ref.ref
    def reverse(self):
        if self.head is None:
            print("linked list is empty")
        else:
            prev = None
            curr = self.head
            while curr is not None:
                nex = curr.ref
                curr.ref = prev
                prev = curr
                curr = nex
            self.head = prev


ll = LinkedList()
ll.add_begin(4)
ll.add_begin(44)
# ll.add_end(55)
ll.add_after(33,55)
ll.display()
ll.add_before(99,4)

ll.add_end(11)
ll.add_end(55)
ll.display()
# ll.delete_begin()
# ll.delete_end()
ll.delete_value(99)
ll.display()
ll.reverse()
ll.display()

File: javaScript/singly_LinkedList.js
class Node {
    constructor(data = null) {
      if (typeof data === null) {
        throw new Error("argument missing at Node(), connot create empty node");
      }
      this.data = data;
      this.next = null;
    }
  }
  
  class LinkedList {
    constructor() {
      this.head = null;
      this.tail = null;
    }
  
    append(val) {
      if (!val) {
        throw new Error("cannot append null node");
      }
      let node = this.head;
      let newNode = new Node(val);
      if (node === null) {
        this.head = newNode;
        this.tail = newNode;
        return this;
      }
  
      this.tail.next = newNode;
      this.tail = newNode;
      return this;
    }
  
    prepend(val = null) {
      if (typeof val === null) {
        throw new Error("cannot prpend null node");
      }
  
      let node = this.head;
      let newNode = new Node(val);
  
      if (node === null) {
        this.head = newNode;
        this.tail = newNode;
        return this;
      }
      this.head = newNode;
      newNode.next = node;
      return this;
    }
  
    insertAfter(data, val) {
      if (data === undefined) {
        throw new Error("argument missing at insertAfter, value is required");
      }
      if (val === undefined) {
        throw new Error(
          "cannot insert null node, argument missing at insertAfter"
        );
      }
      let flag = false;
      let node = this.head;
      let newNode = new Node(val);
      while (node) {
        if (node.data === data) {
          flag = true;
          break;
        }
        node = node.next;
      }
      if (!flag) {
        throw new Error("LinkedList bound out of position");
      }
  
      if (node === this.tail) {
        this.tail = newNode;
      }
      let temp = node.next;
      node.next = newNode;
      newNode.next = temp;
      return this;
    }
  
    insertBefore(data, val) {
      if (data === undefined) {
        throw new Error("argument missing at insertBefore, value is required");
      }
      if (val === undefined) {
        throw new Error(
          "cannot insert null node, argument missing at insertBefore"
        );
      }
      let flag = false;
      let node = this.head;
      let newNode = new Node(val);
      if (this.head.data === data) {
        this.head = newNode;
        newNode.next = node;
        return this;
      }
      while (node) {
        if (node.next.data === data) {
          flag = true;
          break;
        }
        node = node.next;
      }
      if (!flag) {
        throw new Error("LinkedList bound out of position");
      }
      let temp = node.next;
      node.next = newNode;
      newNode.next = temp;
      return this;
    }
  
    insertAt(position, data) {
      if (data === undefined) {
        throw new Error("argument missing at insertAt, value is required");
      }
      if (position === undefined) {
        throw new Error(
          "cannot insertAt null node, argument missing at insertAt"
        );
      }
      let node = this.head;
      let newNode = new Node(data);
      if (position === 0) {
        this.head = newNode;
        newNode.next = node;
        return this;
      }
  
      while (--position) {
        if (node.next !== null) {
          node = node.next;
        } else {
          throw new Error("linkedList out of position");
        }
      }
  
      let temp = node.next;
      node.next = newNode;
      newNode.next = temp;
      return this;
    }
  
    delete(val) {
      if (val === undefined) {
        throw new Error("argument missing at delete");
      }
      let node = this.head;
      if (this.head.data === val) {
        this.head = node.next;
        return this;
      }
      while (node.next) {
        if (node.next.data === val) {
          node.next = node.next.next;
          return this;
        }
        node = node.next;
      }
    }
  
    deleteAfter(val) {
      if (val === undefined) {
        throw new Error("argument missing at deleteAfter");
      }
      let node = this.head;
      while (node.next) {
        if (node.data === val) {
          node.next = node.next.next;
          return this;
        }
        node = node.next;
      }
    }
  
    deleteBefore(data) {
      if (data === undefined) {
        throw new Error("argument missing at deleteBefore, value is required");
      }
  
      if (this.head.data === data) {
        throw new Error("no node before first node");
      }
      let node = this.head;
      if (node.next.data === data) {
        this.head = node.next;
        return this;
      }
      while (node.next) {
        if (node.next?.next?.data === data) {
          node.next = node.next.next;
          return this;
        }
        node = node.next;
      }
    }
  
    deleteAt(position) {
      if (position === undefined) {
        throw new Error(
          "argument missing at deleteAt method, required position to delete element"
        );
      }
      if (position === 0) {
        this.head = this.head.next;
        return this;
      }
      let node = this.head;
      while (--position) {
        if (node.next !== null) {
          node = node.next;
        }
        node = node.next;
      }
      node.next = node.next.next;
      return this;
    }
  
    deleteLast() {
      let node = this.head;
      if (node === null) {
        throw new Error("cannot delete from empty node");
      }
      this.tail = null;
      return this;
    }
  
    deleteFirst() {
      this.head = this.head.next;
      return this;
    }
  
    indexOf(val) {
      if (val === undefined) {
        throw new Error("agrgument missing at indexOf() method, required value");
      }
      let position = -1;
      let node = this.head;
      while (node) {
        ++position;
        if (node.data === val) {
          return position;
        }
        node = node.next;
      }
      return -1;
    }
  
    size() {
      let node = this.head;
      let size = 0;
      while (node) {
        ++size;
        node = node.next;
      }
      return size;
    }
  
    isEmpty() {
      return this.head === null;
    }
  
    clear() {
      this.head = null;
    }
  
    reverse() {
      let previosNode = null;
      let currentNode = this.head;
      let nextNode = null;
      if (this.head === null) return;
      while (currentNode) {
        nextNode = currentNode.next;
        currentNode.next = previosNode;
        previosNode = currentNode;
        currentNode = nextNode;
      }
      this.head = previosNode;
      return this;
    }
  
    printList() {
      let node = this.head;
      process.stdout.write("Head -> ");
      while (node) {
        process.stdout.write(`${node.data} -> `);
        node = node.next;
      }
      process.stdout.write("null");
    }
  }
  
  const list = new LinkedList();
  list.append(1);
  list.append(2);
  list.append(3);
  console.log("size", list.size());
  
  list.reverse();
  list.printList();
  